feat(nuxt): Add "Deployment Provider Setup" (#11486)
getsentry/sentry-docs: docs/platforms/javascript/common/deployment-provider-setup/general.mdx

---
title: General
sidebar_order: 1
description: "Learn about the general deployment setup when using Sentry."
supported:
  - javascript.nuxt
---

<PlatformSection supported={['javascript.nuxt']}>
  Nuxt compiles all your code for the client and server side to ECMAScript modules (ESM) syntax ([learn more about ESM and Nuxt](https://nuxt.com/docs/guide/concepts/esm)).
</PlatformSection>

Adding Sentry on the server-side for a project in ESM syntax is different from adding Sentry to a project in CommonJS (CJS) syntax.
The Sentry initialization code needs to be loaded first when the application starts.
Preloading the Sentry server config file can be done with the Node [`--import`](https://nodejs.org/api/cli.html#--importmodule) CLI flag when starting the application.

Internally, Sentry uses a specific ESM loader hook "[import-in-the-middle/hook.mjs](https://www.npmjs.com/package/import-in-the-middle)" which is

> automatically added when you add the Sentry server-side config file with `--import`. This will automatically instrument all modules in your application.

Reviewer: l: I am not sure if we need to/should include this - this is very much internal stuff that users don't really need to know/understand, and also could possibly change. IMHO it is enough to have the paragraph above, which states that sentry has to be initialized via `--import` ?

## Setup Sentry Server Config

As described above, there are some prerequisites for adding Sentry to a Nuxt project in ESM syntax:

1. The possibility to add the node `--import` CLI flag.

> 2. Knowing the correct path to your Sentry server config file within your deployment environment (to add the path to the `--import` CLI flag).

Reviewer: ```suggestion
2. Knowing the correct path to your Sentry server config file within your deployment environment (as you need to pass it to the `--import` CLI flag).
```

Make sure that this variable is only available during server runtime (not during build time).

### `import-in-the-middle/hook.mjs` is available

Usually, the file `node_modules/import-in-the-middle/hook.mjs` is automatically available when installing dependencies.
However, some deployment providers use custom node bundlers that might not include all `node_modules` in order to save storage.
This is often done by tracing the node files and only including the necessary dependencies. Since Sentry has to include the `hook.mjs` file internally with `module.register()` it might not be discovered by some bundlers and therefore not included.
If you get a warning that the `hook.mjs` file is not found, please report this in [the Sentry JS GitHub repo](https://github.com/getsentry/sentry-javascript/issues) or file an issue at your deployment provider.

Nitro (which is used by Nuxt and SolidStart on the server side) uses [`@vercel/nft`](https://github.com/vercel/nft) during the build. `@vercel/nft` includes files that are added with `module.register()` from version 0.27.4 onwards.

> In case you get a runtime error along the lines `Cannot find module '/node_modules/import-in-the-middle/hook.mjs'`, make sure to add a resolution (yarn) or an override (npm/pnpm) for `@vercel/nft`:

Reviewer: ```suggestion
In case you get a runtime error along the lines `Cannot find module '/node_modules/import-in-the-middle/hook.mjs'`, make sure to add an override (npm/pnpm) or a resolution (yarn) for `@vercel/nft`:
```

we should probably mention npm first (generally)

> ```json {tabTitle:yarn} {filename: package.json}

Reviewer: should we flip these around, npm first, then yarn?

{
  "resolutions": {
    "@vercel/nft": "^0.27.4"
  }
}
```

```json {tabTitle:npm/pnpm} {filename: package.json}
{
  "overrides": {
    "@vercel/nft": "^0.27.4"
  }
}
```

## Setup other environment variables

In case you are using environment variables in your project to set up Sentry don't forget to also add them inside your deployment environment:

```bash
SENTRY_DSN="___PUBLIC_DSN___"
SENTRY_AUTH_TOKEN="___ORG_AUTH_TOKEN___"
```

> 
> 

Reviewer: ```suggestion
```

extra new lines here
